### PR TITLE
NEWS: News update for v1.12.0 rc1

### PR DESCRIPTION
**Pull request by brminich**
## What
- News update for v.1.12.0 release
- Porting v.1.11.x news from v.1.11.x branch

## Why ?
Release preparation

## How ?
Comparing v1.11.x and v1.12.x branches

### CODE REVIEW

**Comment by brminich** [comment]
@Akshay-Venkatesh, can you please add CUDA related features/bug fixes added in ucx 1.12? (if something is missed)

File: NEWS [hunk 1]
```diff
@@ -7,13 +7,177 @@
...
+* Added usage of mpool set for unexpected eager message to reduce memory consumption
+* Added client_id to ucp_worker_create() and ucp_conn_request_query() APIs
+* Added support for modifying UCT and UCS configs by ucp_config_modify() API
+* Added address versioning to correctly preserve wire compatibility
+* Optimized unpacked rkeys memory consumption
+* Added request flag to influence latency vs. bandwidth protocol
+* Added ucp_worker_address_query() API
+* Reduced memory management overhead with new protocols
+* Updated ucp_ep_query() API with getting local and remote addresses
+* Improved performance calculations in new protocols
```
Comment (shamisp) [inline NEWS:32]: in new -> for new ?

File: NEWS [hunk 2]
```diff
@@ -7,13 +7,177 @@
...
+* Added AMO support with new protocols
+* Added stages for rendezvous in new protocols
+* Added put|get_zcopy and pipeline based rendezvous in new protocols
```
Comment (shamisp) [inline NEWS:35]: put|get_zcopy - > put_zcopy/get_zcopy

File: NEWS [hunk 3]
```diff
@@ -7,13 +7,177 @@
...
+* Added support for user-defined alignment in AM
+* Added support for offload tag sync in new protocols
+* Updated ucp_atomic_post() to use NBX flow
+#### UCT
+* Added stub for memory invalidation support in CMA
+* Added IPv6 address support in TCP
+* Added latency estimation to uct_iface_estimate_perf()
+* Adjusted knem and cma overhead cost
+* Added API - uct_iface_is_reachable_v2()
```
Comment (shamisp) [inline NEWS:44]: should go to the top

File: NEWS [hunk 4]
```diff
@@ -7,13 +7,177 @@
...
+* Increased built-in TCP keep-alive interval to 2 seconds
+#### RDMA CORE (IB, ROCE, etc.)
+* Added check for CQ overrun in assert mode
+* Added bitmap usage for releasing detached DCIs
+* Added config for requests ack frequency with DevX
+* Added remote QP info to tx error CQE traces
 #### UCS
-* Added API to Read boot ID value or use machine_guid
+* Added memory pool set data structure
+* Added more accurate topology distance estimation
```
Comment (shamisp) [inline NEWS:53]: Added more accurate -> Improved accuracy of the topology distance estimation

File: NEWS [hunk 5]
```diff
@@ -7,13 +7,177 @@
...
+* Removed a diagnostic message when fuse thread is stopped
+* Added ucs_vsnprintf_safe() which always adds '\0'
+* Added API for a per-process aggregate-sum statistics report
+* Added configurable limit for the memory consumed by rcache
+* Added configuration for VFS(FUSE) thread affinity
+* Added memory limit support to memtrack
+#### CUDA
+* Added option to set cuda_copy bandwidth
+* Added profiling of CUDA runtime function calls
+* Added stub for memory invalidation support
```
Comment (shamisp) [inline NEWS:68]: I'm not sure if we need to list "stub" functions - seems like does not contribute to anything (you have few assurances here).

File: NEWS [hunk 6]
```diff
@@ -7,13 +7,177 @@
...
+* Replaced rand by urand in io_demo
+* More improvements in io_demo
+* Generalized median calculation to support any percentile in ucx_perftest
+#### Tools
+* Added loop-back transport support in ucx_perftest
+* Split ucx_perftest into separate modules
+* Added process placement option for ucx_info
+* Extended parameters correctness check in ucx_perftest
+#### CI
+* Replaced gtest 1.7 with gtest 1.10
```
Comment (shamisp) [inline NEWS:91]: Updated gtest 1.7 to 1.10

File: NEWS [hunk 7]
```diff
@@ -7,13 +7,177 @@
...
+* Increased thresholds for test_ucp_wait_mem
+* Added test for ucx binary compatibility between OS versions
+* Increased test job timeout to 6 hours
+* Reduced testing time under valgrind
+* Added suppressions for glibc and libnl leaks
+* Relaxed performance requirements in perf test
+
+### Bugfixes
+#### Core
+* Fixed simultaneous ep close with ucp_hello_world
```
Comment (shamisp) [inline NEWS:104]: ep -> endpoint

File: NEWS [hunk 8]
```diff
@@ -7,13 +7,177 @@
...
+* Fixes in error flow during rendezvous protocol
+* Multiple fixes in general error flow
+* Fixed fallback to PUT pipeline in rendezvous protocol
+* Reduced default value of keep-alive interval to 20 seconds
+#### UCT
+* Fixed deadlock in TCP
+* Suppressed EHOSTUNREACH error in TCP sockcm
+* Restricted connecting loop-back to other devices in TCP
+#### RDMA CORE (IB, ROCE, etc.)
+* Added pkey_index initialization when creating RC QP with DEVX
```
Comment (shamisp) [inline NEWS:120]: s/Added/Fixed

File: NEWS [hunk 9]
```diff
@@ -7,13 +7,177 @@
...
+* Fixed uninitialized variable access in VFS
+* Changed log level of inotify_init failure to diag
+* Fixed integer overflow in mpool chunk allocation
+#### Packaging
+* Fixed with-fuse arg for RPM build
+#### Documentation
+* Fixes in UCP, UCT, UCS, FAQ and README documentation
+#### Tests
+* Fixed memory leak in io_demo
+* More fixes in io_demo
```
Comment (shamisp) [inline NEWS:147]: remove...sounds duplicated. Maybe replace the both lines with Multiple fixes

**Review by shamisp — CHANGES_REQUESTED** [review]

File: NEWS [hunk 10]
```diff
@@ -7,13 +7,177 @@
...
+* Added initial support for Go language bindings
+* Added memory invalidation on error detection
+* Added threshold for ep connection matching in UCP
+* Added new objects to VFS (md, component, log_level, etc)
+* Added config variable to specify what loadable modules are needed
+* Added sigaction-override config to disable sigaction overriding
 #### UCP
-* Added API for querying UCP library attributes
+* Added new sockaddr private data format
+* Enabled rendezvous and tag sync for all cases with error handling
```
[the remark this answers is not in the document]
Comment (brminich) [reply NEWS:21]: no, previously RNDV and sync were disabled if user used connect to worker address with error handling. Now this restriction is removed

File: NEWS (hunk 5 above)
Comment (brminich) [reply NEWS:68]: this is important change, because it allows these transports to be selected for RNDV protocol.
Remove?

**Comment by brminich** [comment]
@shamisp, fixed

**Comment by Akshay-Venkatesh** [comment]
> @Akshay-Venkatesh, can you please add CUDA related features/bug fixes added in ucx 1.12? (if something is missed)

@brminich I'm copying this from Yossi's talk this week. These were the main additions: 
- added global memtype cache​ to allow UCT transports to query memory attributes
- added capability to select cuda stream based on source and destination memory type​ required for device-memory-based pipelining
- Auto-register cuda whole allocation​s to avoid repeated registration costs
- Select Cuda-IPC capabilities based on NVLINK topology​ to prefer writes vs reads for specific platforms using NVML
- Generalized rendezvous fragment pool per mem-type device

There were some recent bug-fixes but they're part of master and not v1.12.x

File: NEWS (hunk 5 above)
Comment (shamisp) [reply NEWS:69]: I would rephrase this:  Added stub for memory invalidation support ignorer to enable CUDA transport selection for rendezvous protocols

Comment (dmitrygx) [reply NEWS:69]: wdyt `Added stub for memory invalidation support ignorer to enable CUDA transport selection for rendezvous protocols in case of error handling enabled`?

Comment (shamisp) [reply NEWS:69]: 👍

Comment (shamisp) [reply NEWS:69]: Actually,  I would make it a bit shorter `Added stub for memory invalidation support to enable CUDA-IPC transport selection for rendezvous protocols`

File: NEWS [hunk 11]
```diff
@@ -7,13 +7,177 @@
 ##
 #
 
-## Current
+## 1.12.0 RC1 (December 2, 2021)
 ### Features:
+#### Core
+* Added initial support for Go language bindings
```
Comment (yosefe) [inline NEWS:13]: remove "initial"

File: NEWS [hunk 12]
```diff
@@ -7,13 +7,177 @@
...
+* Added memory invalidation on error detection
```
Comment (yosefe) [inline NEWS:14]: Bugfix : Fixed invalid remote memory access after connection error

File: NEWS [hunk 13]
```diff
@@ -7,13 +7,177 @@
...
+* Added threshold for endpoint connection matching in UCP
```
Comment (yosefe) [inline NEWS:15]: is this a bugfix?

File: NEWS [hunk 14]
```diff
@@ -7,13 +7,177 @@
...
+* Added new objects to VFS (md, component, log_level, etc)
+* Added configuration variable to specify what loadable modules are needed
```
Comment (yosefe) [inline NEWS:17]: Added configuration variable to specify which loadable modules are allowed

File: NEWS [hunk 15]
```diff
@@ -7,13 +7,177 @@
...
+* Added sigaction-override configuration to disable sigaction overriding
```
Comment (yosefe) [inline NEWS:18]: Added build-time configuration to disable sigaction overriding

File: NEWS [hunk 16]
```diff
@@ -7,13 +7,177 @@
...
 #### UCP
-* Added API for querying UCP library attributes
+* Added address versioning to correctly preserve wire compatibility since v1.11.0
+* Added new sockaddr private data format
```
Comment (yosefe) [inline NEWS:21]: Added new client/server connection establishment packet header format

File: NEWS (hunk 10 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply NEWS:22]: Enabled rendezvous and tag sync protocols when error handling is enabled on the endpoint

File: NEWS [hunk 17]
```diff
@@ -7,13 +7,177 @@
...
+* Enabled rendezvous and tag sync for all cases with error handling
+* Added iov zcopy support to RMA operations
+* Added usage of mpool set for unexpected eager message to reduce memory consumption
```
Comment (yosefe) [inline NEWS:24]: Reduced memory usage of unexpected messages by fitting receive buffer size to packet size

File: NEWS [hunk 18]
```diff
@@ -7,13 +7,177 @@
...
+* Added support for modifying UCT and UCS configs by ucp_config_modify() API
+* Optimized unpacked rkeys memory consumption
+* Added request flag to influence latency vs. bandwidth protocol
```
Comment (yosefe) [inline NEWS:27]: Added request flag to influence latency vs. bandwidth optimization

File: NEWS [hunk 19]
```diff
@@ -7,13 +7,177 @@
...
+* Reduced memory management overhead with new protocols
+* Improved performance calculations for new protocols
+* Added AMO support with new protocols
```
Comment (yosefe) [inline NEWS:30]: Added AMO support with GPU memory target using new protocols

File: NEWS [hunk 20]
```diff
@@ -7,13 +7,177 @@
...
+* Added stages for rendezvous in new protocols
```
Comment (yosefe) [inline NEWS:31]: can remove this

File: NEWS [hunk 21]
```diff
@@ -7,13 +7,177 @@
...
+* Added put_zcopy, get_zcopy and pipeline based rendezvous in new protocols
+* Added support for user-defined alignment in AM
```
Comment (yosefe) [inline NEWS:33]: AM -> active messages

File: NEWS (hunk 5 above)
Comment (yosefe) [reply NEWS:69]: i would remove this altogether, it's part of memory invalidation err flows fix

File: NEWS [hunk 22]
```diff
@@ -7,13 +7,177 @@
...
+* Extended statistics in io_demo
+* Improved logging in io_demo
+* Replaced rand by urand in io_demo
+* More improvements in io_demo
+* Generalized median calculation to support any percentile in ucx_perftest
+#### Tools
+* Added loop-back transport support in ucx_perftest
+* Split ucx_perftest into separate modules
+* Added process placement option for ucx_info
+* Extended parameters correctness check in ucx_perftest
```
Comment (yosefe) [inline NEWS:90]: Add:
Added support for GPU memory RMA and atomics

**Comment by yosefe** [comment]
@petro-rudenko can you pls check didn't miss any news for Java

File: NEWS (hunk 11 above)
Comment (shamisp) [reply NEWS:13]: @yosefe Do we position as a preview or we can claim release. If this is release, I think we should move it to the section of its own, like Java

**Comment by petro-rudenko** [comment]
JUCX: 
- Set worker id and query it from the connection request.
- ucp_listener_reject functionality.
- Support setting port 0 for UcpListener to bind on a free port.

File: NEWS (hunk 11 above)
Comment (yosefe) [reply NEWS:13]: how about
Added beta-level support for Go language bindings

Comment (shamisp) [reply NEWS:13]: +1

File: NEWS (hunk 16 above)
Comment (yosefe) [reply NEWS:19]: missed?

File: NEWS (hunk 17 above)
Comment (yosefe) [reply NEWS:22]: missed?

**Review by yosefe — COMMENTED** [review]
seems few comments missed from prev review

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@shamisp can you pls take a look?

File: NEWS [hunk 23]
```diff
@@ -7,13 +7,183 @@
...
+* Added memory pool set data structure
+* Improved accuracy of the topology distance estimation
+* Added thread-safe put to ptr_map
+* Added prints of leaked callbacks from the callback queue
+* Added new ptr_array API for bulk allocation
+* Added ucs_string_buffer_append_flags() for string buffer
+* Added ucs_ffs32()
+* Removed a diagnostic message when fuse thread is stopped
+* Added ucs_vsnprintf_safe() which always adds '\0'
+* Added API for a per-process aggregate-sum statistics report
```
Comment (shamisp) [inline NEWS:58]: Please move API changes to the top of UCS section

**Review by shamisp — CHANGES_REQUESTED** [review]

File: NEWS [hunk 24]
```diff
@@ -7,13 +7,183 @@
...
+* Optimized unpacked rkeys memory consumption
+* Added request flag to influence latency vs. bandwidth protocol
+* Reduced memory management overhead with new protocols
+* Improved performance calculations for new protocols
+* Added AMO support with GPU memory target using new protocols
+* Added put_zcopy, get_zcopy and pipeline based rendezvous in new protocols
+* Added support for user-defined alignment in Active Messages
+* Added support for offload tag sync in new protocols
+* Updated ucp_atomic_post() to use NBX flow
+##### API
```
[the remark this answers is not in the document]
Comment (brminich) [reply NEWS:33]: but this is UCP API changes, it has nested level (extra `#`)

[the remark this answers is not in the document]
Comment (brminich) [reply NEWS:33]: or you mean move it right after `####UCP`?

File: NEWS [hunk 25]
```diff
@@ -7,13 +7,183 @@
...
+* Added selection of CUDA-IPC capabilities based on NVLINK topology
+  (to prefer writes vs reads for specific platforms using NVML)
+* Added option to set cuda_copy bandwidth
+* Added profiling of CUDA runtime function calls
+#### Java
+* Added ucp_listener_reject functionality
+* Added support for setting worker id and querying it from the connection request
+* Added support to bind on a free port in UcpListener
+#### Packaging
+* Added cmake support
```
[the remark this answers is not in the document]
Comment (brminich) [reply NEWS:76]: afair #7096

File: NEWS (hunk 24 above)
[the remark this answers is not in the document]
Comment (shamisp) [reply NEWS:33]: Yes.

File: NEWS (hunk 25 above)
[the remark this answers is not in the document]
Comment (shamisp) [reply NEWS:76]: Added cmake config files for better integration with external cmake based projects

File: NEWS (hunk 24 above)
[the remark this answers is not in the document]
Comment (brminich) [reply NEWS]: moved to the top, but removed `#####API` line, because:
1. We do not have such section for other parts (UCT, UCS), so it is consistent with other parts - that API changes are just on the top
2. Otherwise would need to introduce some other caption which would separate API and plain features

File: NEWS [hunk 26]
```diff
@@ -7,13 +7,183 @@
...
+* Added bitmap usage for releasing detached DCIs
+* Added configuration for requests ack frequency with DevX
+* Added remote QP info to tx error CQE traces
 #### UCS
-* Added API to Read boot ID value or use machine_guid
+* Added memory pool set data structure
+* Improved accuracy of the topology distance estimation
+* Added thread-safe put to ptr_map
+* Added prints of leaked callbacks from the callback queue
+* Added new ptr_array API for bulk allocation
```
[the remark this answers is not in the document]
Comment (brminich) [reply NEWS:50]: moved

File: NEWS (hunk 23 above)
Comment (brminich) [reply NEWS]: moved

File: NEWS (hunk 25 above)
[the remark this answers is not in the document]
Comment (brminich) [reply NEWS]: updated

**Comment by brminich** [comment]
bot:pipe:retest

**Review by shamisp — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

File: NEWS [hunk 27]
```diff
@@ -7,13 +7,182 @@
...
+* Added memory limit support to memtrack
+#### CUDA
+* Added global memtype cache to allow UCT transports to query memory attributes
+* Auto-register cuda whole allocations to avoid repeated registration costs
+* Added capability to select CUDA stream based on source and destination memory type
+  (required for device memory based pipelining)
+* Added selection of CUDA-IPC capabilities based on NVLINK topology
+  (to prefer writes vs reads for specific platforms using NVML)
+* Added option to set cuda_copy bandwidth
+* Added profiling of CUDA runtime function calls
```
Comment (bureddy) [inline NEWS:69]: Can you add the following in CUDA section
 "Added option to limit GPUDirectRDMA size in rendezvous protocol"

Comment (brminich) [reply NEWS:69]: added

**Comment by brminich** [comment]
@shamisp, @yosefe let's keep it open to be able to update if something else gets into v1.12

**Comment by brminich** [comment]
@tonycurtis, can you please take a look?

File: NEWS [hunk 28]
```diff
@@ -7,13 +7,183 @@
...
+#### Core
+* Added beta-level support for Go language bindings
+* Added new objects to VFS (md, component, log_level, etc.)
+* Added configuration variable to specify which loadable modules are allowed
+* Added build-time configuration to disable sigaction overriding
 #### UCP
-* Added API for querying UCP library attributes
+* Added client_id to ucp_worker_create() and ucp_conn_request_query() APIs
+* Added ucp_worker_address_query() API
+* Updated ucp_ep_query() API with getting local and remote addresses
```
Comment (tonycurtis) [inline NEWS:20]: for getting?

File: NEWS [hunk 29]
```diff
@@ -7,13 +7,183 @@
...
+* Added address versioning to correctly preserve wire compatibility starting from the version v1.11.0
```
Comment (tonycurtis) [inline NEWS:21]: remove "the", and the "v" in front of the number seems redundant.

File: NEWS [hunk 30]
```diff
@@ -7,13 +7,183 @@
...
+* Added thread-safe put to ptr_map
+* Improved accuracy of the topology distance estimation
+* Added prints of leaked callbacks from the callback queue
+* Removed a diagnostic message when fuse thread is stopped
+* Added configurable limit for the memory consumed by rcache
+* Added configuration for VFS(FUSE) thread affinity
+* Added memory limit support to memtrack
+#### CUDA
+* Added global memtype cache to allow UCT transports to query memory attributes
+* Auto-register cuda whole allocations to avoid repeated registration costs
```
Comment (tonycurtis) [inline NEWS:63]: cuda -> CUDA

File: NEWS [hunk 31]
```diff
@@ -7,13 +7,183 @@
...
+* Added capability to select CUDA stream based on source and destination memory type
+  (required for device memory based pipelining)
+* Added selection of CUDA-IPC capabilities based on NVLINK topology
+  (to prefer writes vs reads for specific platforms using NVML)
```
Comment (tonycurtis) [inline NEWS:67]: vs.    (period on end)

File: NEWS [hunk 32]
```diff
@@ -7,13 +7,183 @@
...
+* Fixes in switching transports during client/server connection setup
+* Fixes in transport-level error reporting
+
+## 1.11.1 (August 31, 2021)
+### Features:
+#### UCS
+* Added API to read boot ID value or use machine_guid
+
 ### Bugfixes:
+* Fixes in Cuda memory hooks
```
Comment (tonycurtis) [inline NEWS:177]: Cuda -> CUDA

File: NEWS [hunk 33]
```diff
@@ -68,7 +238,7 @@
 #### RDMA CORE (IB, ROCE, etc.)
 * Added report of QP info in case of completion with error
 * Refactored of FC send operations
```
Comment (tonycurtis) [inline NEWS:240]: ah, not added by this issue, but looks like a missing word here

**Review by tonycurtis — CHANGES_REQUESTED** [review]

**Comment by shamisp** [comment]
> @shamisp, @yosefe let's keep it open to be able to update if something else gets into v1.12

makes sense. You also want to update Authors files (in separate PR)

**Comment by brminich** [comment]
> > @shamisp, @yosefe let's keep it open to be able to update if something else gets into v1.12
> 
> makes sense. You also want to update Authors files (in separate PR)

Probably I was wrong. Would be good to merge it to not block rc1.
@tonycurtis, your comments applied. If you do not have any new comments I'd squash the changes

**Review by tonycurtis — APPROVED** [review]

**Comment by shamisp** [comment]
> > > @shamisp, @yosefe let's keep it open to be able to update if something else gets into v1.12
> > 
> > 
> > makes sense. You also want to update Authors files (in separate PR)
> 
> Probably I was wrong. Would be good to merge it to not block rc1. @tonycurtis, your comments applied. If you do not have any new comments I'd squash the changes

You and @tonycurtis  have to decide. In the past we have tried to push NEWs on the fist RC but sometimes those were delayed.

**Comment by tonycurtis** [comment]
> On Dec 9, 2021, at 5:31 PM, Pavel Shamis (Pasha) ***@***.***> wrote:
> 
> 
> @shamisp <https://github.com/shamisp>, @yosefe <https://github.com/yosefe> let's keep it open to be able to update if something else gets into v1.12
> 
> makes sense. You also want to update Authors files (in separate PR)
> 
> Probably I was wrong. Would be good to merge it to not block rc1. @tonycurtis <https://github.com/tonycurtis>, your comments applied. If you do not have any new comments I'd squash the changes
> 
> You and @tonycurtis <https://github.com/tonycurtis> have to decide. In the past we have tried to push NEWs on the fist RC but sometimes those were delayed.
> 

I could convince myself either way.  Keep a running update of NEWS; or accumulate a separate PR until the release.

I’d go for the latter, if pushed.  I think people understand an RC is not a release, so the meta-documentation can wait for a final update.

Tony

**Comment by brminich** [comment]
Imo, better have actual NEWs for rc release, because users may/will want to try it and check its content

**Comment by shamisp** [comment]
@brminich agree. If you happen to have ready to go before RC I don't see a good reason not to include. I have seen few some linux distress using our RC.

**Comment by brminich** [comment]
@yosefe, @tonycurtis, so I'm going to merge it unless you have some objections

**Comment by tonycurtis** [comment]
> On Dec 10, 2021, at 9:20 AM, Mikhail Brinskiy ***@***.***> wrote:
> 
> 
> @yosefe <https://github.com/yosefe>, @tonycurtis <https://github.com/tonycurtis>, so I'm going to merge it unless you have some objections
> 

No, please go ahead.
Tony
